### DISCUSSION TITLE
Re-implement ride construction window as plugin

### DISCUSSION THREAD

**Discussion by spacek531** [OpenRCT2/OpenRCT2#16690] [Feature requests]
As a NEDesigns user, there are a few difficult-to-access ride features that have no access without plugins. Currently, I have to switch between ride types to access track elements not available to a given ride type - it would be helpful if I could build all track elements regardless of if the ride supports them. Currently, there is no way to change a track element's draw mode except with plugins that - as far as I know - do not exist. Not yet implemented is the supports feature, which I would likewise benefit having easy access to.

A re-implementation of the ride construction window as a plugin, one that can be forked from, can expose new and useful ride construction features without having to modify the ingame ride construction window for every use case.

**Comment by 733737** [comment]
are you saying to replace the ride construction window c++ code entirely with javascript code?

**Comment by spacek531** [comment]
Plugin API that will likely be required:

~~TrackElementDescriptors~~ Framework already implemented, just need to implement a few more properties
RideTypeDescriptors or at least the enabled track pieces
Vehicle object flags (to support new behaviour for @X123M3-256's wiegand ride)

**Reply by X123M3-256** [reply]
> Vehicle object flags (to support new behaviour for @X123M3-256's wiegand ride)

Do you mean the ride type flags? That ride doesn't have any vehicle object flag set that would affect the ride construction UI.

**Reply by spacek531** [reply]
It's an RTD flag? I thought it was a vehicle object flag.

**Comment by spacek531** [comment]
I remember now that some users do not have scripting, so it is not feasible to remove the native implementation entirely. It still raises the question of what features need native representation and which can be represented only by plugins. For example the only way to access TrackElement RideType is via plugin. Should it be required that all track elements are buildable by the native UI? @X123M3-256's proposed multi-length slope transitions come to mind as being difficult to represent with the existing paradigm.

**Reply by X123M3-256** [reply]
I think that non-cheat elements should be buildable by the native UI, but the arbitrary ride type changing is a hack so I think it's fine if it's only accessible by plugin (some rides may benefit from having multiple track types as a proper feature, but I'd do that with dinghy slide style buttons, which doesn't really work for selecting an arbitrary track type).
